File: mergesort-1.hs
import System.Random
import Control.Monad
import Data.List
import System.CPUTime

---Merge Sort in Haskell---

-- Merge sort implementation
mergeSort :: (Ord a) => [a] -> [a]
mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs = merge (mergeSort firstHalf) (mergeSort secondHalf)
  where
    -- Split the input list into two equal halves
    (firstHalf, secondHalf) = splitAt (length xs `div` 2) xs

    -- Merge two sorted lists into a single sorted list
    merge :: (Ord a) => [a] -> [a] -> [a]
    merge xs [] = xs
    merge [] ys = ys
    merge (x:xs) (y:ys)
      | x <= y = x : merge xs (y:ys)
      | otherwise = y : merge (x:xs) ys

-- Generates a list of n random numbers
generateRandomNumbers :: Int -> IO [Int]
generateRandomNumbers n
  | n > 0 = do
    gen <- newStdGen
    return $ take n (randomRs (1, 1000) gen)
  | otherwise = return []

-- Measures the execution time of the given sorting function with the provided input
measureExecutionTime :: ([Int] -> [Int]) -> String -> [Int] -> IO ()
measureExecutionTime sortFn listType input = do
  startTime <- getCPUTime
  let sortedNumbers = sortFn input
  endTime <- getCPUTime
  print $ "Sorting " ++ show (length input) ++ " " ++ listType ++ " numbers took " ++ show (fromIntegral (endTime - startTime) / 10^12) ++ " seconds."

main :: IO ()
main = do
  let n1 = 1000
  randomNumbers <- generateRandomNumbers n1
  measureExecutionTime mergeSort "random" randomNumbers

  let n2 = 1000000
  let sortedList = [1..n2]
  let reversedList = reverse sortedList

  measureExecutionTime mergeSort "sorted" sortedList
  measureExecutionTime mergeSort "reversed" reversedList

File: mergesort1.py
import random
import time

def merge_sort(arr):
    """
    Merge Sort In Python

    Sorts an array using the merge sort algorithm.
    :param arr: The array to be sorted.
    :return: The sorted array.
    """
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]
    left = merge_sort(left)
    right = merge_sort(right)
    return merge(left, right)

def merge(left, right):
    """
    Merges two sorted arrays into a single sorted array.

    :param left: The left sorted array.
    :param right: The right sorted array.
    :return: The merged sorted array.
    """
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result += left[i:]
    result += right[j:]
    return result

def generate_random_numbers(n):
    """
    Generates a list of n random integers between 1 and 1000.

    :param n: The number of integers to generate.
    :return: The list of random integers.
    """
    return [random.randint(1, 1000) for _ in range(n)]

def measure_execution_time(sort_fn, list_type, input):
    """
    Measures the execution time of a sorting function.

    :param sort_fn: The sorting function to be measured.
    :param list_type: The type of list being sorted (e.g., "random", "sorted", "reversed").
    :param input: The list to be sorted.
    """
    start_time = time.perf_counter()
    sorted_numbers = sort_fn(input)
    end_time = time.perf_counter()
    elapsed_time = end_time - start_time
    print(f"'Sorting {len(input)} {list_type} numbers took {elapsed_time:.3e} seconds.'")
    if sorted_numbers != sorted(input):
        print("'Error: list not sorted correctly!'")


random_numbers = generate_random_numbers(1000)
measure_execution_time(merge_sort, "random", random_numbers)

sorted_list = list(range(1, 1000001))
reversed_list = sorted_list[::-1]

measure_execution_time(merge_sort, "sorted", sorted_list)
measure_execution_time(merge_sort, "reversed", reversed_list)
